File: python/agentbay/exceptions.py
class AgentBayError(Exception):
    """Base exception for all AgentBay SDK errors."""

    def __init__(self, message=None, *args, **kwargs):
        if message is None:
            message = self.__class__.__name__
        super().__init__(message, *args)
        self.extra = kwargs


class AuthenticationError(AgentBayError):
    """Raised when there is an authentication error."""

    def __init__(self, message="Authentication failed", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class APIError(AgentBayError):
    """Raised when there is an error with the API."""

    def __init__(self, message="API error", status_code=None, *args, **kwargs):
        super().__init__(message, *args, **kwargs)
        self.status_code = status_code


class FileError(AgentBayError):
    """Raised for errors related to file operations."""

    def __init__(self, message="File operation error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class CommandError(AgentBayError):
    """Raised for errors related to command execution."""

    def __init__(self, message="Command execution error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class SessionError(AgentBayError):
    """Raised for errors related to session operations."""

    def __init__(self, message="Session error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class OssError(AgentBayError):
    """Raised for errors related to OSS operations."""

    def __init__(self, message="OSS operation error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class ApplicationError(AgentBayError):
    """Raised for errors related to application operations."""

    def __init__(self, message="Application operation error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)


class UIError(AgentBayError):
    """Raised for errors related to UI operations."""

    def __init__(self, message="UI operation error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)

class BrowserError(AgentBayError):
    """Raised when there is an error with the browser."""

    def __init__(self, message="Browser error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)

class AgentError(AgentBayError):
    """Raised for errors related to Agent actions."""

    def __init__(self, message="Agent action error", *args, **kwargs):
        super().__init__(message, *args, **kwargs)

File: typescript/src/exceptions.ts
/**
 * Base exception for all AgentBay SDK errors.
 */
export class AgentBayError extends Error {
  extra: Record<string, any>;

  constructor(message?: string, extra: Record<string, any> = {}) {
    const errorMessage = message || "AgentBayError";
    super(errorMessage);
    this.name = "AgentBayError";
    this.extra = extra;
    Object.setPrototypeOf(this, AgentBayError.prototype);
  }
}

/**
 * Raised when there is an authentication error.
 */
export class AuthenticationError extends AgentBayError {
  constructor(
    message = "Authentication failed",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "AuthenticationError";
    Object.setPrototypeOf(this, AuthenticationError.prototype);
  }
}

/**
 * Raised when there is an error with the API.
 */
export class APIError extends AgentBayError {
  statusCode?: number;

  constructor(
    message = "API error",
    statusCode?: number,
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "APIError";
    this.statusCode = statusCode;
    Object.setPrototypeOf(this, APIError.prototype);
  }
}

/**
 * Raised for errors related to file operations.
 */
export class FileError extends AgentBayError {
  constructor(
    message = "File operation error",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "FileError";
    Object.setPrototypeOf(this, FileError.prototype);
  }
}

/**
 * Raised for errors related to command execution.
 */
export class CommandError extends AgentBayError {
  constructor(
    message = "Command execution error",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "CommandError";
    Object.setPrototypeOf(this, CommandError.prototype);
  }
}

/**
 * Raised for errors related to session operations.
 */
export class SessionError extends AgentBayError {
  constructor(message = "Session error", extra: Record<string, any> = {}) {
    super(message, extra);
    this.name = "SessionError";
    Object.setPrototypeOf(this, SessionError.prototype);
  }
}

/**
 * Raised for errors related to OSS operations.
 */
export class OssError extends AgentBayError {
  constructor(
    message = "OSS operation error",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "OssError";
    Object.setPrototypeOf(this, OssError.prototype);
  }
}

/**
 * Raised for errors related to application operations.
 */
export class ApplicationError extends AgentBayError {
  constructor(
    message = "Application operation error",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "ApplicationError";
    Object.setPrototypeOf(this, ApplicationError.prototype);
  }
}

/**
 * Raised for errors related to UI operations.
 */
export class UIError extends AgentBayError {
  constructor(message = "UI operation error", extra: Record<string, any> = {}) {
    super(message, extra);
    this.name = "UIError";
    Object.setPrototypeOf(this, UIError.prototype);
  }
}

/**
 * Raised for errors related to browser operations.
 */
export class BrowserError extends AgentBayError {
  constructor(
    message = "Browser operation error",
    extra: Record<string, any> = {}
  ) {
    super(message, extra);
    this.name = "BrowserError";
    Object.setPrototypeOf(this, BrowserError.prototype);
  }
}
